pkg/asset/handler/presign_upload.go

```diff
@@ -0,0 +1,167 @@
+package handler
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz/policy"
+	"github.com/skygeario/skygear-server/pkg/core/cloudstorage"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	"github.com/skygeario/skygear-server/pkg/core/inject"
+	"github.com/skygeario/skygear-server/pkg/core/server"
+	"github.com/skygeario/skygear-server/pkg/core/skyerr"
+	"github.com/skygeario/skygear-server/pkg/core/validation"
+)
+
+func AttachPresignUploadHandler(
+	server *server.Server,
+	dependencyMap inject.DependencyMap,
+) *server.Server {
+	server.Handle("/presign_upload", &PresignUploadHandlerFactory{
+		dependencyMap,
+	}).Methods("OPTIONS", "POST")
+	return server
+}
+
+type PresignUploadHandlerFactory struct {
+	DependencyMap inject.DependencyMap
+}
+
+func (f *PresignUploadHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &PresignUploadHandler{}
+	inject.DefaultRequestInject(h, f.DependencyMap, request)
+	return h.RequireAuthz(h, h)
+}
+
+// @JSONSchema
+const PresignUploadResponseSchema = `
+{
+	"$id": "#PresignUploadResponse",
+	"type": "object",
+	"properties": {
+		"result": {
+			"type": "object",
+			"properties": {
+				"asset_name": { "type": "string" },
+				"url": { "type": "string" },
+				"method": { "type": "string" },
+				"headers": {
+					"type": "array",
+					"items": {
+						"type": "object",
+						"properties": {
+							"name": { "type": "string" },
+							"value": { "type": "string" }
+						},
+						"required": ["name", "value"]
+					}
+				}
+			},
+			"required": ["asset_name", "url", "method", "headers"]
+		}
+	}
+}
+`
+
+/*
+	@Operation POST /presign_upload - Presign an upload request.
+		Presign an upload request.
+
+		@SecurityRequirement access_key
+		@SecurityRequirement access_token
+
+		@RequestBody
+			@JSONSchema {PresignUploadRequest}
+
+		@Response 200
+			@JSONSchema {PresignUploadResponse}
+*/
+type PresignUploadHandler struct {
+	RequireAuthz         handler.RequireAuthz  `dependency:"RequireAuthz"`
+	CloudStorageProvider cloudstorage.Provider `dependency:"CloudStorageProvider"`
+	Validator            *validation.Validator `dependency:"Validator"`
+}
+
+func (h *PresignUploadHandler) ProvideAuthzPolicy() authz.Policy {
+	return policy.AnyOf(
+		authz.PolicyFunc(policy.RequireMasterKey),
+		policy.AllOf(
+			authz.PolicyFunc(policy.DenyNoAccessKey),
+			authz.PolicyFunc(policy.RequireAuthenticated),
+			authz.PolicyFunc(policy.DenyDisabledUser),
+		),
+	)
+}
+
+// @JSONSchema
+const PresignUploadRequestSchema = `
+{
+	"$id": "#PresignUploadRequest",
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"prefix": {
+			"type": "string",
+			"pattern": "^[^\\x00\\\\/:*'<>|]*$"
```

Comment:
Do you have any reference on this regex?

Reply:
I forgot. Let me revise it to match https://en.wikipedia.org/wiki/Filename#Reserved_characters_and_words

Reply:
Let me revise it to `a-z` `A-Z` `0-9` `_-.`